### PR TITLE
Add shapes module to p32 and update other crates to latest version

### PR DESCRIPTION
**Pull request by bubka42**

### CODE REVIEW

File: p32/src/shapes.rs [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+    const NAME: &'static str;
+
+    fn area(&self) -> f32;
+    fn perimeter(&self) -> f32;
+    fn scale(&mut self, factor: f32);
+
+    fn area_to_perimeter(&self) -> f32 {
+        match self.perimeter() {
+            0.0 => 0.0,
+            _ => self.area() / self.perimeter(),
```
Comment (newpavlov) [inline p32/src/shapes.rs]: This method should not special-case the zero case. Instead, `Point` should overwrite default impl for this method.

File: p32/src/shapes.rs [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+        };
+    }
+}
+
+pub enum Return {
+    First,
+    Second,
+}
+
+pub fn bigger_area_to_perimeter<'a>(first: &'a impl Shape, second: &'a impl Shape) -> Return {
```
Comment (newpavlov) [inline p32/src/shapes.rs]: This function should accept two shape _slices_  and return _reference_ to the biggest shape.

Comment (bubka42) [reply p32/src/shapes.rs]: How should I use enum to return reference to a trait? Should I use DynamicShape wrapper?

Comment (bubka42) [reply p32/src/shapes.rs]: I will come back to this one later.

Comment (newpavlov) [reply p32/src/shapes.rs]: You should make the enum itself generic. Each variant should contain a reference to a shape of different type.

Comment (bubka42) [reply p32/src/shapes.rs]: Got it. And should each slice have multiple shapes in them? And should I return the shape with the biggest area-to-perimeter among all the shapes in the two slices? Or just the slice with the larger total-area-to-total-perimeter?

Comment (newpavlov) [reply p32/src/shapes.rs]: Each slice should contain shapes of one type. The return result should contain reference to one shape with the biggest ratio.

Comment (bubka42) [reply p32/src/shapes.rs]: Can you check now?

File: p32/src/shapes.rs [hunk 3]
```diff
@@ -176,17 +176,40 @@ impl Shape for DynamicShape {
...
+    let smax = second
+        .iter()
+        .map(|shape| shape.area_to_perimeter())
+        .reduce(f32::max)
+        .unwrap();
+    if fmax > smax {
+        let (fmax_idx, _) = first
+            .iter()
+            .enumerate()
+            .find(|(_, shape)| shape.area_to_perimeter() == fmax)
```
Comment (newpavlov) [inline p32/src/shapes.rs]: This iteration is not necessary. In the first and second iteration  you can return `(f32, &T)` and `(f32, &U)` accordingly.

File: p32/src/shapes.rs [hunk 4]
```diff
@@ -176,17 +176,40 @@ impl Shape for DynamicShape {
...
-// }
+pub enum Return<'a, T, U> {
+    First(&'a T),
+    Second(&'a U),
+}
+
+pub fn bigger_area_to_perimeter<'a>(
+    first: &'a [impl Shape],
+    second: &'a [impl Shape],
+) -> Return<'a, impl Shape, impl Shape> {
```
Comment (newpavlov) [inline p32/src/shapes.rs]: In this case it's better to use `fn biggest_area_to_perimeter<'a, 'b, T: Shape, U: Shape>(first: &[T], second: &[U]) -> Return<'a, 'b, T, U>`.

File: p32/src/shapes.rs [hunk 5]
```diff
@@ -176,17 +176,40 @@ impl Shape for DynamicShape {
...
+            .unwrap();
+        println!("{:#?}", first[fmax_idx]);
+        Return::First(&first[fmax_idx])
+    } else {
+        let (smax_idx, _) = second
+            .iter()
+            .enumerate()
+            .find(|(_, shape)| shape.area_to_perimeter() == smax)
+            .unwrap();
+        println!("{:#?}", second[smax_idx]);
```
Comment (newpavlov) [inline p32/src/shapes.rs]: Remove these `println`s and write a test instead.

**Review by newpavlov — APPROVED** [review]

**Pull request merged**